sys/userlib/src/lib.rs

```diff
@@ -1284,71 +1284,161 @@ pub unsafe extern "C" fn _start() -> ! {
+            //   construction).
+            // - The dest pointer must be writable for the given number of
+            //   bytes, which we ensure the same way except by using
+            //   `remaining`.
+            // - Both pointers must be properly aligned, which is ensured by
+            //   getting them from slices.
+            // - The affected regions starting at the source and dest pointers
+            //   must not overlap, which we can prove trivially since they're
+            //   coming from a `&[u8]` and a `&mut [u8]`, which cannot overlap
+            //   due to `&mut` aliasing requirements.
+            unsafe {
+                core::ptr::copy_nonoverlapping(
+                    strbytes.as_ptr(),
+                    remaining.as_mut_ptr(),
+                    to_write,
+                );
+            }
+            // Finally, update our `pos` to record the number of bytes written.
+            //
+            // We use a wrapping add here to avoid integer overflow checks that
+            // could recursively panic. However, we also need to maintain our
+            // type invariant that `pos` does not exceed `buf.len()`. We can be
+            // sure of this here because:
+            //
+            // - `remaining.len()` is exactly equal to `buf.len() - pos`.
+            // - `to_write` is no larger than `remaining.len()`.
+            // - Therefore `pos + to_write <= buf.len()`.
+            //
+            // While this code is not `unsafe`, violating this invariant would
+            // break the `unsafe` code above, hence the size of this comment on
+            // what is otherwise a single instruction.
+            self.pos = self.pos.wrapping_add(to_write);
+
             Ok(())
         }
     }
 
-    // Okay. Now we start the actual panicking process. Allocate buffer:
-    let mut pw = PrefixWrite([0; BUFSIZE], 0);
-    // Generate message:
+    // We declare a single static panic buffer per task, to ensure the memory is
+    // available.
+    static mut PANIC_BUFFER: [u8; BUFSIZE] = [0; BUFSIZE];
+
+    // Okay. Now we start the actual panicking process.
+    //
+    // Safety: this is unsafe because we're getting a reference to a static mut,
+    // and the compiler can't be sure we're not aliasing or racing. We can be
+    // sure that this reference won't be aliased elsewhere in the program,
+    // because we've lexically confined it to this block.
+    //
+    // However, it is possible to produce an alias if the panic handler is
+    // called reentrantly. This can only happen if the code in the panic handler
+    // itself panics, which is what we're working very hard to prevent here.
+    let panic_buffer = unsafe { &mut PANIC_BUFFER };
+
+    // Whew! Time to write the darn message.
+    //
+    // Note that if we provided a different value of `pos` here we could destroy
+    // PrefixWrite's type invariant, so, don't do that.
```

Comment:
ah yes so i guess given that there is literal access to the pos, that's why i don't think of it as a type invariant. however, *really* isolating this stuff would make the code more convoluted for little gain, so im not advocating that exactly

Reply:
Right -- strictly I'd need to move this into a submodule and protect its internals. So far I couldn't be bothered, so, you do have to read the whole function to verify the invariant. Open to better terms!